src/journal/superblock.rs

```diff
@@ -0,0 +1,260 @@
+    pub struct JournalIncompatibleFeatures: u32 {
+        const BLOCK_REVOCATIONS = 0x1;
+        const IS_64BIT = 0x2;
+        const ASYNC_COMMITS = 0x4;
+        const CHECKSUM_V2 = 0x8;
+        const CHECKSUM_V3 = 0x10;
+        const FAST_COMMITS = 0x20;
+    }
+}
+
+#[cfg(all(test, feature = "std"))]
+mod tests {
+    use super::*;
+
+    fn write_u32be(bytes: &mut [u8], offset: usize, value: u32) {
+        bytes[offset..offset + 4].copy_from_slice(&value.to_be_bytes());
+    }
+
+    fn create_test_superblock() -> Vec<u8> {
+        let mut block = vec![0; 1024];
+        // Set magic.
+        write_u32be(&mut block, 0, JournalBlockHeader::MAGIC);
+        // Set superblock type.
+        write_u32be(&mut block, 4, 4);
+        // Set block size.
+        write_u32be(&mut block, SUPERBLOCK_BLOCKSIZE_OFFSET, 4096);
+        // Set sequence.
+        write_u32be(&mut block, SUPERBLOCK_SEQUENCE_OFFSET, 123);
+        // Set start block.
+        write_u32be(&mut block, SUPERBLOCK_START_OFFSET, 456);
+        // Set features.
+        write_u32be(&mut block, SUPERBLOCK_FEATURE_INCOMPAT_OFFSET, 0x12);
+        // Set UUID.
+        block[SUPERBLOCK_UUID_OFFSET..SUPERBLOCK_UUID_OFFSET + 16]
+            .copy_from_slice(&[0xab; 16]);
+        // Set checksum type.
+        block[SUPERBLOCK_CHECKSUM_TYPE_OFFSET] = CHECKSUM_TYPE_CRC32C;
+        // Set checksum.
+        write_u32be(&mut block, SUPERBLOCK_CHECKSUM_OFFSET, 0x78a2_c32b);
+        block
+    }
+
+    #[test]
+    fn test_journal_superblock_read_success() {
+        let block = create_test_superblock();
+        assert_eq!(
+            JournalSuperblock::read_bytes(&block).unwrap(),
+            JournalSuperblock {
+                block_size: 4096,
+                sequence: 123,
+                start_block: 456,
+                uuid: Uuid([0xab; 16]),
+            }
+        );
+    }
+
+    #[test]
+    fn test_journal_superblock_invalid_magic() {
+        let mut block = create_test_superblock();
+        // Override magic in the block header.
+        write_u32be(&mut block, 0, 0);
+        assert_eq!(
+            JournalSuperblock::read_bytes(&block).unwrap_err(),
+            CorruptKind::JournalMagic
+        );
+    }
+
+    #[test]
+    fn test_journal_superblock_unsupported_type() {
+        let mut block = create_test_superblock();
+        // Override type in the block header.
+        write_u32be(&mut block, 4, 0);
+        assert_eq!(
+            JournalSuperblock::read_bytes(&block).unwrap_err(),
+            Incompatible::JournalSuperblockType(0)
+        );
+    }
+
+    #[test]
+    fn test_journal_superblock_missing_required_features() {
+        let mut block = create_test_superblock();
+        write_u32be(&mut block, SUPERBLOCK_FEATURE_INCOMPAT_OFFSET, 0x10);
+        assert_eq!(
+            JournalSuperblock::read_bytes(&block).unwrap_err(),
+            Incompatible::JournalIncompatibleFeatures(0x10),
+        );
+    }
```

Comment:
nit: could optionally add a test for unsupported incompat features?

Reply:
done

Reply:
or, not done. good thing you asked for this, as I'm realizing I updated the test incompletely and it still passes, revealing that `bitflags` doesn't behave quite like I thought it did.

Reply:
OK now it is fixed properly.